manager/state/raft/raft.go

```diff
@@ -1178,21 +1239,64 @@ func (n *Node) sendToMember(members map[uint64]*membership.Member, m raftpb.Mess
 		}
 		n.ReportUnreachable(m.To)
 
+		lastSeenHost := n.cluster.LastSeenHost(m.To)
+		if lastSeenHost != "" {
+			// Check if address has changed
+			officialHost, officialPort, _ := net.SplitHostPort(conn.Addr)
+			if officialHost != lastSeenHost {
+				reconnectAddr := net.JoinHostPort(lastSeenHost, officialPort)
+				n.handleAddressChange(conn, reconnectAddr)
+				return
+			}
+		}
+
 		// Bounce the connection
 		newConn, err := n.ConnectToMember(conn.Addr, 0)
 		if err != nil {
 			n.Config.Logger.Errorf("could connect to member ID %x at %s: %v", m.To, conn.Addr, err)
-		} else {
-			err = n.cluster.ReplaceMemberConnection(m.To, conn, newConn)
-			if err != nil {
-				newConn.Conn.Close()
-			}
+			return
+		}
+		if n.cluster.ReplaceMemberConnection(m.To, conn, newConn, conn.Addr, false) != nil {
+			newConn.Conn.Close()
 		}
 	} else if m.Type == raftpb.MsgSnap {
 		n.ReportSnapshot(m.To, raft.SnapshotFinish)
 	}
 }
 
+func (n *Node) handleAddressChange(member *membership.Member, reconnectAddr string) {
+	newConn, err := n.ConnectToMember(reconnectAddr, 0)
+	if err != nil {
+		n.Config.Logger.Errorf("could connect to member ID %x at observed address %s: %v", member.RaftID, reconnectAddr, err)
+		return
+	}
+
+	healthClient := api.NewHealthClient(member.Conn)
+
+	healthCtx, cancelHealth := context.WithTimeout(n.Ctx, time.Duration(n.Config.ElectionTick)*time.Second)
```

Comment:
btw, it bothered me for a long time, that `TickInterval` is configurable, but not exposed inside `Node` object, it should be something like `n.tickInterval` not `time.Second` I think.